File: src/com/nadmm/airports/wx/TafService.java
/*
 * FlightIntel for Pilots
 *
 * Copyright 2012 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 */

package com.nadmm.airports.wx;

import java.io.File;

import android.content.Intent;
import android.text.format.DateUtils;

import com.nadmm.airports.R;
import com.nadmm.airports.utils.UiUtils;

public class TafService extends NoaaService {

    private final String TAF_IMAGE_NAME = "taf_%s.gif";
    private final String TAF_TEXT_QUERY = "dataSource=tafs&requestType=retrieve"
            +"&format=xml&compression=gzip&hoursBeforeNow=%d&mostRecent=true&stationString=%s";
    private final String TAF_IMAGE_PATH = "/tools/weatherproducts/tafs/default/"
            + "loadImage/region/%s/product/prevail/zoom/%s";

    private static final long TAF_CACHE_MAX_AGE = 2*DateUtils.HOUR_IN_MILLIS;

    protected TafParser mParser;

    public TafService() {
        super( "taf", TAF_CACHE_MAX_AGE );
        mParser = new TafParser();
    }

    @Override
    protected void onHandleIntent( Intent intent ) {
        String action = intent.getAction();
        if ( action.equals( ACTION_GET_TAF ) ) {
            String type = intent.getStringExtra( TYPE );
            if ( type.equals( TYPE_TEXT ) ) {
                // Get request parameters
                String stationId = intent.getStringExtra( STATION_ID );
                int hours = intent.getIntExtra( HOURS_BEFORE, 6 );
                boolean cacheOnly = intent.getBooleanExtra( CACHE_ONLY, false );
                boolean forceRefresh = intent.getBooleanExtra( FORCE_REFRESH, false );

                File xml = getDataFile( "TAF_"+stationId+".xml" );

                if ( forceRefresh || ( !cacheOnly && !xml.exists() ) ) {
                    fetchTaf( stationId, hours, xml );
                }

                Taf taf = new Taf();

                if ( xml.exists() ) {
                    taf.stationId = stationId;
                    mParser.parse( xml, taf );
                }

                // Broadcast the result
                Intent result = makeIntent( action, type );
                result.putExtra( STATION_ID, stationId );
                result.putExtra( RESULT, taf );
                sendBroadcast( result );
            } else if ( type.equals( TYPE_IMAGE ) ) {
                String code = intent.getStringExtra( IMAGE_CODE );
                String imageName = String.format( TAF_IMAGE_NAME, code );
                File image = getDataFile( imageName );
                if ( !image.exists() ) {
                    try {
                        boolean hiRes = getResources().getBoolean( R.bool.WxHiResImages );
                        String path = String.format( TAF_IMAGE_PATH, code,
                                hiRes? "true" : "false" );
                        fetchFromNoaa( path, null, image, false );
                    } catch ( Exception e ) {
                        UiUtils.showToast( this, "Unable to fetch TAF image: "+e.getMessage() );
                    }
                }

                // Broadcast the result
                Intent result = makeIntent( action, type );
                result.putExtra( IMAGE_CODE, code );
                if ( image.exists() ) {
                    result.putExtra( RESULT, image.getAbsolutePath() );
                }
                sendBroadcast( result );
            }
        }
    }

    protected boolean fetchTaf( String stationId, int hours, File xml ) {
        try {
            String query = String.format( TAF_TEXT_QUERY, hours, stationId );
            return fetchFromNoaa( query, xml, true );
        } catch ( Exception e ) {
            UiUtils.showToast( this, "Unable to fetch TAF: "+e.getMessage() );
        }
        return false;
    }

}

File: src/main/java/com/nadmm/airports/wx/TafService.kt
/*
 * FlightIntel for Pilots
 *
 * Copyright 2012-2025 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nadmm.airports.wx

import android.content.Context
import android.content.Intent
import android.text.format.DateUtils
import android.util.Log
import androidx.core.os.bundleOf
import com.nadmm.airports.utils.UiUtils.showToast
import kotlinx.coroutines.launch
import java.io.File

class TafService : NoaaService("taf", CACHE_MAX_AGE) {

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        intent?.let {
            val action = intent.action

            serviceScope.launch {
                if (action == ACTION_GET_TAF) {
                    val type = intent.getStringExtra(TYPE)
                    if (type == TYPE_TEXT) {
                        getTafText(intent)
                    }
                }
            }
        }

        return START_NOT_STICKY
    }

    private suspend fun getTafText(intent: Intent) {
        // Get request parameters
        val stationId = intent.getStringExtra(STATION_ID) ?: return
        val forceRefresh = intent.getBooleanExtra(FORCE_REFRESH, false)

        val cachedFile = wxCache.getCachedFile(stationId)
        if (forceRefresh) {
            // If force refresh is requested, delete the cached files
            cachedFile.delete()
        }

        Log.d(TAG, "getTafText: stationId=$stationId, forceRefresh=$forceRefresh")

        if (!cachedFile.exists()) {
            var xmlFile: File? = null
            try {
                xmlFile = wxCache.createTempFile()
                val query = ("dataSource=tafs&requestType=retrieve"
                        + "&hoursBeforeNow=${HOURS_BEFORE}&mostRecentForEachStation=constraint"
                        + "&format=xml&stationString=${stationId}")
                fetchFromNoaa(query, xmlFile)
                val parser = TafParser()
                val taf = parser.parse(xmlFile)
                wxCache.serializeObject(taf, stationId)
            } catch (e: Exception) {
                showToast(this, "Unable to fetch TAF: " + e.message)
            } finally {
                xmlFile?.delete()
            }
        }

        val taf = wxCache.deserializeObject<Taf>(stationId) ?: Taf()
        val result = bundleOf(
            ACTION to ACTION_GET_TAF,
            TYPE to TYPE_TEXT,
            RESULT to taf
        )
        Events.post(result)
    }

    companion object {
        private val TAG = TafService::class.java.simpleName
        private const val CACHE_MAX_AGE = 30 * DateUtils.MINUTE_IN_MILLIS
        private const val HOURS_BEFORE = 6

        const val TAF_RADIUS = 25

        // Helper function to start the service
        fun startService(context: Context, stationId: String, refresh: Boolean) {
            val intent = Intent(context, TafService::class.java).apply {
                action = ACTION_GET_TAF
                putExtra(STATION_ID, stationId)
                putExtra(TYPE, TYPE_TEXT)
                putExtra(FORCE_REFRESH, refresh)
            }
            context.startService(intent)
        }
    }
}
